### PR TITLE
Send to route

### PR DESCRIPTION
**Pull request by araspitzu**
Implement sending a payment to a pre-computed route, the route input format is compatible with the output of the `/findroute` API. As the route input format is a simple list of nodeId(s) it is mandatory that all the channels involved in the pre-computed route are public, thus sending to a private channel is not supported in this API. The behavior is very similar to other payments API, in a fire-and-forget fashion the endpoint immediately returns a UUID of the payment attempt and more information can then be retrieved via the websocket or payments APIs, the payment will not be retried with different routes (maxAttempts=1).

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/api/FormParamExtractors.scala [hunk 1]
```diff
@@ -57,4 +57,10 @@ object FormParamExtractors {
     Timeout(str.toInt.seconds)
   }
 
+  implicit val pubkeyListUnmarshaller: Unmarshaller[String, List[PublicKey]] = Unmarshaller.strict { str =>
+    serialization.read[List[String]](str).map { el =>
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/FormParamExtractors.scala]: Does that parse URL-encoded json?

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentInitiator.scala [hunk 2]
```diff
@@ -19,18 +19,18 @@ package fr.acinq.eclair.payment
...
+import fr.acinq.eclair.payment.PaymentLifecycle.{GenericSendPayment, SendPayment, SendPaymentToRoute}
 
 /**
   * Created by PM on 29/08/2016.
   */
 class PaymentInitiator(nodeParams: NodeParams, router: ActorRef, register: ActorRef) extends Actor with ActorLogging {
 
   override def receive: Receive = {
-    case c: SendPayment =>
+    case c:GenericSendPayment =>
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentInitiator.scala]:
```suggestion
    case c: GenericSendPayment =>
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentInitiator.scala [hunk 3]
```diff
@@ -19,18 +19,18 @@ package fr.acinq.eclair.payment
...
       val paymentId = UUID.randomUUID()
       val payFsm = context.actorOf(PaymentLifecycle.props(nodeParams, paymentId, router, register))
-      payFsm forward c
+      payFsm.forward(c)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentInitiator.scala]: Previous syntaxt was a convention we use elsewhere (to be close to `actor ! msg`)

```suggestion
      payFsm forward c
```

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 4]
```diff
@@ -67,6 +67,7 @@ case class RouteRequest(source: PublicKey,
                         ignoreChannels: Set[ChannelDesc] = Set.empty,
                         routeParams: Option[RouteParams] = None)
 
+case class PartialRouteRequest(hops:Seq[PublicKey])
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: The name is misleading, it makes it seem like we have a partial route and we need the rest. How about (credits @sstone):

```suggestion
case class FinalizeRoute(hops:Seq[PublicKey])
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala [hunk 5]
```diff
@@ -45,6 +45,12 @@ class PaymentLifecycle(nodeParams: NodeParams, id: UUID, router: ActorRef, regis
   startWith(WAITING_FOR_REQUEST, WaitingForRequest)
 
   when(WAITING_FOR_REQUEST) {
+    case Event(SendPaymentToRoute(amountMsat, paymentHash, hops, finalCltvExpiry), WaitingForRequest) =>
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]: nit: that's not consistent with the handler style below (I found that argument extraction is less maintenable, everytime you add an arg they don't compile).

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentInitiator.scala [hunk 6]
```diff
@@ -19,18 +19,18 @@ package fr.acinq.eclair.payment
 import java.util.UUID
 import akka.actor.{Actor, ActorLogging, ActorRef, Props}
 import fr.acinq.eclair.NodeParams
-import fr.acinq.eclair.payment.PaymentLifecycle.SendPayment
+import fr.acinq.eclair.payment.PaymentLifecycle.{GenericSendPayment, SendPayment, SendPaymentToRoute}
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentInitiator.scala]: Unused imports

```suggestion
import fr.acinq.eclair.payment.PaymentLifecycle.GenericSendPayment
```

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 4 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: I came up with the name because the route here is not complete, I like the suggestion though!

Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: Addressed in 44acb29

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala (hunk 5 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]: I'm okay with this proposed change even though in general i like the compiler to do the heavy lifting and error-catching, but in the case of the extractor pattern the compile time errors are typically not very useful, addressed in 52e4938.

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentInitiator.scala (hunk 6 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentInitiator.scala]: Addressed in 3584079

File: eclair-core/src/main/scala/fr/acinq/eclair/api/FormParamExtractors.scala (hunk 1 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/FormParamExtractors.scala]: There is no URL-encoding here, this only parses the form data parameter (as json array of strings) into a list of public keys.

**Comment by araspitzu** [comment]
> Wouldn't that be possible to add a `goingThroughNodes` attributes to the existing `SendPayment`? Or have a `Hint` parameter that can be `NoHint`, `GoesThrough(nodeId: PublicKey)` or `StaticRoute(nodeIds: Seq[PublicKey])`

Yes that was an option but adding "hints" to `SendPayment` feels wrong because it would have a double meaning, on one hand it would refer to the BOLT11 hints found on the payment request, on the other we would store an entire static route in there. This would lead to some branching code where we take different action depending on the value of the hints, and i believe it's not the best approach in term of readability. The two use cases are captured by the duo SendPayment/SendPaymentToRoute.

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 7]
```diff
@@ -410,6 +411,13 @@ class Router(nodeParams: NodeParams, watcher: ActorRef, initialized: Option[Prom
       sender ! d
       stay
 
+    case Event(FinalizeRoute(partialHops), d) =>
+      // split into sublists [(a,b),(b,c), ...] then get the edges between each of those pair, then select the largest edge between them
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]:
```suggestion
      // split into sublists [(a,b),(b,c), ...] then get the edges between each of those pairs, then select the largest edge between them
```

File: eclair-core/src/test/scala/fr/acinq/eclair/router/RouterSpec.scala [hunk 8]
```diff
@@ -130,10 +130,10 @@ class RouterSpec extends BaseRouterSpec {
   test("handle bad signature for NodeAnnouncement") { fixture =>
     import fixture._
     val sender = TestProbe()
-    val buggy_ann_a = ann_a.copy(signature = ann_b.signature, timestamp = ann_a.timestamp + 1)
-    sender.send(router, PeerRoutingMessage(null, remoteNodeId, buggy_ann_a))
-    sender.expectMsg(TransportHandler.ReadAck(buggy_ann_a))
-    sender.expectMsg(InvalidSignature(buggy_ann_a))
+    val buggy_ann_b = ann_b.copy(signature = ann_c.signature, timestamp = ann_b.timestamp + 1)
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/RouterSpec.scala]: Why did you need to change that?

Comment (araspitzu) [reply eclair-core/src/test/scala/fr/acinq/eclair/router/RouterSpec.scala]: Good catch that was some leftover from development, in 3aa79f7 i reverted the changes to RouterSpec and BaseRouterSpec (along with updating the copyright header).

**Comment by araspitzu** [comment]
TODO: update doc

**Comment by pm47** [comment]
Just realized there is no API-level test, ideally it would test both input formats (json and comma-separated):
- `["nodeid1", "nodeid2", ... ]`
- `nodeid1,nodeid2`
